source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -0,0 +1,149 @@
+    m_left.read(stream);
+    m_right.read(stream);
+    m_jugemIndex = stream.read_u8();
+    m_type = stream.read_s8();
+    m_prevPt = stream.read_u8();
+    m_nextPt = stream.read_u8();
+}
+
+bool MapdataCheckPoint::isPlayerFlagged(s32 /* playerIdx */) const {
+    return m_flag;
+}
+
+void MapdataCheckPoint::setPlayerFlags(s32 /* playerIdx */) {
+    m_flag = true;
+}
+
+void MapdataCheckPoint::resetFlags() {
+    m_flag = false;
+}
+
+const EGG::Vector2f &MapdataCheckPoint::left() const {
+    return m_left;
+}
+
+const EGG::Vector2f &MapdataCheckPoint::right() const {
+    return m_right;
+}
+
+u8 MapdataCheckPoint::jugemIndex() const {
+    return m_jugemIndex;
+}
+
+s8 MapdataCheckPoint::type() const {
+    return m_type;
+}
+
+bool MapdataCheckPoint::isNormalCheckpoint() const {
+    return m_type == NORMAL_CHECKPOINT;
+}
+
+bool MapdataCheckPoint::isFinishLine() const {
+    return m_type == FINISH_LINE;
+}
+
+u16 MapdataCheckPoint::nextCount() const {
+    return m_nextCount;
+}
+
+u16 MapdataCheckPoint::prevCount() const {
+    return m_prevCount;
+}
+
+u16 MapdataCheckPoint::id() const {
+    return m_id;
+}
+
+MapdataCheckPoint *MapdataCheckPoint::prevPoint(s32 i) const {
+    return m_prevPoints[i];
+}
+
+MapdataCheckPoint *MapdataCheckPoint::nextPoint(s32 i) const {
+    return m_nextPoints[i].checkpoint;
+}
+
+const LinkedCheckpoint &MapdataCheckPoint::nextLinked(s32 i) const {
+    return m_nextPoints[i];
+}
+
+/// @brief computes @ref m_prevKcpId for each checkpoint.
+/// @details @ref isPlayerFlagged indicates that a checkpoint has been visited, to prevent infinite
+/// recursion.
+/// @addr{0x80515A6C}
+void MapdataCheckPoint::linkPrevKcpIds(u8 prevKcpId) {
+    m_prevKcpId = isNormalCheckpoint() ? prevKcpId : m_type;
+
+    setPlayerFlags(0);
+    for (size_t i = 0; i < nextCount(); i++) {
+        MapdataCheckPoint *next = nextPoint(i);
+        if (!next->isPlayerFlagged(0)) {
+            next->linkPrevKcpIds(m_prevKcpId);
+        }
+    }
+}
+
+/// @brief find finish line and last key checkpoint indexes; also initCheckpointLinks for all
+/// checkpoints
+/// @addr{Inlined in 0x80515244} fake. not real. it's not in the base game. in the base
+/// game, it's inlined into `init()`. i, kooshnoo, split it out because i wanted to.
+// I'm not sure if this should be split; it feels like a separate function,
+//  but idk if i can somehow prove nintendo wrote it as a separate function,
+//  and i don't know if we even care what nintendo extracted to a separate function and what they
+//  inlined anyways. would the PR reviewer kindly please advise, thank you
+void MapdataCheckPointAccessor::findFinishAndLastKcp() {
+    s8 lastKcpType = -1;
+    s16 finishLineCheckpointId = -1;
+
+    for (size_t ckptId = 0; ckptId < size(); ckptId++) {
+        MapdataCheckPoint *checkpoint = get(ckptId);
+        // @todo implement initCheckpointLinks
+        // checkpoint->initCheckpointLinks(*this, ckptId);
```

Comment:
Does this need to be left here?

Reply:
this function is the topic of a future PR. if you leave out the comments, the code makes no sense:
```cpp
MapdataCheckPoint *checkpoint = get(ckptId);
checkpoint = get(ckptId);
```

should i remove both comments, the top 'todo' one only, or the bottom placeholder one only?

Reply:
Nah you can leave this in then